Looks like this broke paradise

```
java.lang.AssertionError: assertion failed: 
  (class InteropIdentity,copy$default$1)
     while compiling: /home/travis/build/scala/scala/test/macro-annot/src/test/scala/run/InteropCaseSynthesis.scala
        during phase: typer
     library version: version 2.13.0-20180528-033852-61322fc
    compiler version: version 2.13.0-20180528-033852-61322fc
  reconstructed args: -classpath /home/travis/build/scala/scala/test/macro-annot/target/test-classes:/home/travis/build/scala/scala/test/macro-annot/target/classes:/home/travis/build/scala/scala/build/quick/classes/reflect:/home/travis/build/scala/scala/build/quick/classes/compiler:/home/travis/build/scala/scala/build/quick/classes/repl:/home/travis/build/scala/scala/build/quick/classes/interactive:/home/travis/build/scala/scala/build/quick/classes/repl-frontend:/home/travis/build/scala/scala/build/quick/classes/scaladoc:/home/travis/.ivy2/cache/org.scala-lang.modules/scala-asm/bundles/scala-asm-6.0.0-scala-1.jar:/home/travis/.ivy2/cache/jline/jline/jars/jline-2.14.6.jar:/home/travis/.ivy2/cache/junit/junit/jars/junit-4.11.jar:/home/travis/.ivy2/cache/org.hamcrest/hamcrest-core/jars/hamcrest-core-1.3.jar:/home/travis/.ivy2/cache/com.novocode/junit-interface/jars/junit-interface-0.11.jar:/home/travis/.ivy2/cache/org.scala-sbt/test-interface/jars/test-interface-1.0.jar -Ywarn-unused-import -bootclasspath /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/resources.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/rt.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/sunrsasign.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/jsse.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/jce.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/charsets.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/jfr.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/classes:/home/travis/build/scala/scala/build/quick/classes/library -Xfatal-warnings -Ywarn-unused:imports -Ymacro-annotations
  last tree to typer: This(class InteropIdentity)
       tree position: line 6 of /home/travis/build/scala/scala/test/macro-annot/src/test/scala/run/InteropCaseSynthesis.scala
            tree tpe: InteropIdentity.this.type
              symbol: case class InteropIdentity
   symbol definition: case class InteropIdentity extends Product with Serializable (a ClassSymbol)
      symbol package: <empty>
       symbol owners: class InteropIdentity
           call site: method copy in class InteropIdentity in package <empty>
== Source file context for tree position ==
     3 import org.junit.runners._
     4 import Assert._
     5 
     6 @identity case class InteropIdentity(x: Int)
     7 @placebo case class InteropPlacebo(x: Int)
     8 
     9 @RunWith(classOf[JUnit4])
	at scala.reflect.internal.SymbolTable.throwAssertionError(SymbolTable.scala:163)
	at scala.tools.nsc.typechecker.Namers$Namer.$anonfun$addDefaultGetters$10(Namers.scala:1572)
	at scala.tools.nsc.typechecker.Namers$Namer$DefaultMethodInOwningScope.addGetter(Namers.scala:1648)
	at scala.tools.nsc.typechecker.Namers$Namer.$anonfun$addDefaultGetters$8(Namers.scala:1541)
	at scala.tools.nsc.typechecker.Namers$Namer.$anonfun$addDefaultGetters$6(Namers.scala:1512)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:126)
	at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:72)
	at scala.tools.nsc.typechecker.Namers$Namer.addDefaultGetters(Namers.scala:1508)
	at scala.tools.nsc.typechecker.Namers$Namer.methodSig(Namers.scala:1395)
	at scala.tools.nsc.typechecker.Namers$Namer.memberSig(Namers.scala:1860)
	at scala.tools.nsc.typechecker.Namers$Namer.typeSig(Namers.scala:1826)
	at scala.tools.nsc.typechecker.Namers$Namer$MonoTypeCompleter.completeImpl(Namers.scala:848)
	at scala.tools.nsc.typechecker.Namers$LockingTypeCompleter.complete(Namers.scala:2023)
	at scala.tools.nsc.typechecker.Namers$LockingTypeCompleter.complete$(Namers.scala:2021)
	at scala.tools.nsc.typechecker.Namers$TypeCompleterBase.complete(Namers.scala:2016)
	at scala.tools.nsc.typechecker.Namers$CompleterWrapper.complete(Namers.scala:2072)
	at scala.tools.nsc.typechecker.Namers$Namer$$anon$3.complete(Namers.scala:612)
	at scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1527)
	at scala.reflect.internal.Symbols$Symbol.initialize(Symbols.scala:1675)
	at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typedStats$6(Typers.scala:3198)
	at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typedStats$6$adapted(Typers.scala:3196)
	at scala.Option$WithFilter.foreach(Option.scala:270)
	at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typedStats$4(Typers.scala:3196)
	at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typedStats$4$adapted(Typers.scala:3194)
	at scala.reflect.internal.Scopes$Scope.foreach(Scopes.scala:409)
	at scala.tools.nsc.typechecker.Typers$Typer.addSynthetics$1(Typers.scala:3194)
	at scala.tools.nsc.typechecker.Typers$Typer.typedStats(Typers.scala:3256)
```

--- a/src/compiler/scala/tools/nsc/typechecker/MacroAnnotationNamers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/MacroAnnotationNamers.scala
@@ -272,8 +272,13 @@ trait MacroAnnotationNamers { self: Analyzer =>
             if (isEnumConstant(tree))
               tree.symbol setInfo ConstantType(Constant(tree.symbol))
           case tree @ DefDef(_, nme.CONSTRUCTOR, _, _, _, _) =>
+            if (mexists(tree.vparamss)(_.mods.hasDefault))
+              enterDefaultGetters(tree.symbol, tree, tree.vparamss, tree.tparams)
             sym setInfo completerOf(tree)
           case tree @ DefDef(mods, name, tparams, _, _, _) =>
+            if (mexists(tree.vparamss)(_.mods.hasDefault))
+              enterDefaultGetters(tree.symbol, tree, tree.vparamss, tree.tparams)
+
             val completer =
               if (sym hasFlag SYNTHETIC) {
                 if (name == nme.copy) copyMethodCompleter(tree)

--- a/src/compiler/scala/tools/nsc/typechecker/Namers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Namers.scala
@@ -745,20 +745,24 @@ trait Namers extends MethodSynthesis {
 
     def enterTypeDef(tree: TypeDef) = assignAndEnterFinishedSymbol(tree)
 
-    def enterDefDef(tree: DefDef): Unit = tree match {
-      case DefDef(_, nme.CONSTRUCTOR, _, _, _, _) =>
-        assignAndEnterFinishedSymbol(tree)
-      case DefDef(mods, name, _, _, _, _) =>
-        val sym = enterInScope(assignMemberSymbol(tree))
+    def enterDefDef(tree: DefDef): Unit = {
+      tree match {
+        case DefDef(_, nme.CONSTRUCTOR, _, _, _, _) =>
+          assignAndEnterFinishedSymbol(tree)
+        case DefDef(mods, name, _, _, _, _) =>
+          val sym = enterInScope(assignMemberSymbol(tree))
 
-        val completer =
-          if (sym hasFlag SYNTHETIC) {
-            if (name == nme.copy) copyMethodCompleter(tree)
-            else if (sym hasFlag CASE) applyUnapplyMethodCompleter(tree, context)
-            else completerOf(tree)
-          } else completerOf(tree)
+          val completer =
+            if (sym hasFlag SYNTHETIC) {
+              if (name == nme.copy) copyMethodCompleter(tree)
+              else if (sym hasFlag CASE) applyUnapplyMethodCompleter(tree, context)
+              else completerOf(tree)
+            } else completerOf(tree)
 
-        sym setInfo completer
+          sym setInfo completer
+      }
+      if (mexists(tree.vparamss)(_.mods.hasDefault))
+        enterDefaultGetters(tree.symbol, tree, tree.vparamss, tree.tparams)
     }
 
     def enterClassDef(tree: ClassDef): Unit = {
@@ -1189,6 +1193,12 @@ trait Namers extends MethodSynthesis {
       val module = clazz.sourceModule
       for (cda <- module.attachments.get[ConstructorDefaultsAttachment]) {
         debuglog(s"Storing the template namer in the ConstructorDefaultsAttachment of ${module.debugLocationString}.")
+        if (cda.defaults.nonEmpty) {
+          for (sym <- cda.defaults) {
+            decls.enter(sym)
+          }
+          cda.defaults.clear()
+        }
         cda.companionModuleClassNamer = templateNamer
       }
       val classTp = ClassInfoType(parents, decls, clazz)
@@ -1420,6 +1430,42 @@ trait Namers extends MethodSynthesis {
     }
 
     /**
+     * For every default argument, insert a method symbol computing that default
+     */
+    def enterDefaultGetters(meth: Symbol, ddef: DefDef, vparamss: List[List[ValDef]], tparams: List[TypeDef]) {
+      val methOwner  = meth.owner
+      val search = DefaultGetterNamerSearch(context, meth, initCompanionModule = false)
+      var posCounter = 1
+
+      mforeach(vparamss){(vparam) =>
+        // true if the corresponding parameter of the base class has a default argument
+        if (vparam.mods.hasDefault) {
+          val name = nme.defaultGetterName(meth.name, posCounter)
+
+          search.createAndEnter { owner: Symbol =>
+            methOwner.resetFlag(INTERFACE) // there's a concrete member now
+            val default = owner.newMethodSymbol(name, vparam.pos, paramFlagsToDefaultGetter(meth.flags))
+            default.setPrivateWithin(meth.privateWithin)
+            default.referenced = meth
+            default.setInfo(ErrorType)
+            if (meth.name == nme.apply && meth.hasAllFlags(CASE | SYNTHETIC)) {
+              val att = meth.attachments.get[CaseApplyDefaultGetters].getOrElse({
+                val a = new CaseApplyDefaultGetters()
+                meth.updateAttachment(a)
+                a
+              })
+              att.defaultGetters += default
+            }
+            if (default.owner.isTerm)
+              saveDefaultGetter(meth, default)
+            default
+          }
+        }
+        posCounter += 1
+      }
+    }
+
+    /**
      * For every default argument, insert a method computing that default
      *
      * Also adds the "override" and "defaultparam" (for inherited defaults) flags
@@ -1491,24 +1537,6 @@ trait Namers extends MethodSynthesis {
             val defVparamss = mmap(rvparamss.take(previous.length)){ rvp =>
               copyValDef(rvp)(mods = rvp.mods &~ DEFAULTPARAM, rhs = EmptyTree)
             }
-            val defaultGetterSym = search.createAndEnter { owner: Symbol =>
-              methOwner.resetFlag(INTERFACE) // there's a concrete member now
-              val default = owner.newMethodSymbol(name, vparam.pos, paramFlagsToDefaultGetter(meth.flags))
-              default.setPrivateWithin(meth.privateWithin)
-              if (meth.name == nme.apply && meth.hasAllFlags(CASE | SYNTHETIC)) {
-                val att = meth.attachments.get[CaseApplyDefaultGetters].getOrElse({
-                  val a = new CaseApplyDefaultGetters()
-                  meth.updateAttachment(a)
-                  a
-                })
-                att.defaultGetters += default
-              }
-              if (default.owner.isTerm)
-                saveDefaultGetter(meth, default)
-              default
-            }
-            if (defaultGetterSym == NoSymbol) return
-
             search.addGetter(rtparams) {
               (parentNamer: Namer, defTparams: List[TypeDef]) =>
                 val defTpt =
@@ -1536,6 +1564,11 @@ trait Namers extends MethodSynthesis {
                 val defaultTree = atPos(vparam.pos.focus) {
                   DefDef(Modifiers(paramFlagsToDefaultGetter(meth.flags), ddef.mods.privateWithin) | oflag, name, defTparams, defVparamss, defTpt, defRhs)
                 }
+                def referencesThis(sym: Symbol) = sym match {
+                  case term: TermSymbol => term.referenced == meth
+                  case _ => false
+                }
+                val defaultGetterSym = parentNamer.context.scope.lookup(name).filter(referencesThis)
                 assert(defaultGetterSym != NoSymbol, (parentNamer.owner, name))
                 defaultTree.setSymbol(defaultGetterSym)
                 defaultGetterSym.setInfo(parentNamer.completerOf(defaultTree))
@@ -1562,7 +1595,7 @@ trait Namers extends MethodSynthesis {
     private abstract class DefaultGetterNamerSearch {
       def addGetter(rtparams0: List[TypeDef])(create: (Namer, List[TypeDef]) => Tree)
 
-      def createAndEnter(f: Symbol => Symbol): Symbol
+      def createAndEnter(f: Symbol => Symbol): Unit
     }
     private class DefaultGetterInCompanion(c: Context, meth: Symbol, initCompanionModule: Boolean) extends DefaultGetterNamerSearch {
       private val module = companionSymbolOf(meth.owner, context)
@@ -1570,14 +1603,19 @@ trait Namers extends MethodSynthesis {
       private val cda: Option[ConstructorDefaultsAttachment] = module.attachments.get[ConstructorDefaultsAttachment]
       private val moduleNamer = cda.flatMap(x => Option(x.companionModuleClassNamer))
 
-      def createAndEnter(f: Symbol => Symbol): Symbol = {
+      def createAndEnter(f: Symbol => Symbol): Unit = {
         val default = f(module.moduleClass)
         moduleNamer match {
           case Some(namer) =>
             namer.enterInScope(default)
           case None =>
-            // ignore error to fix #3649 (prevent crash in erroneous source code)
-            NoSymbol
+            cda match {
+              case Some(attachment) =>
+                // defer entry until the companion module body it type completed
+                attachment.defaults += default
+              case None =>
+              // ignore error to fix #3649 (prevent crash in erroneous source code)
+            }
         }
       }
       def addGetter(rtparams0: List[TypeDef])(create: (Namer, List[TypeDef]) => Tree): Unit = {
@@ -1603,7 +1641,7 @@ trait Namers extends MethodSynthesis {
         assert(ctx != NoContext, meth)
         newNamer(ctx)
       }
-      def createAndEnter(f: Symbol => Symbol): Symbol = {
+      def createAndEnter(f: Symbol => Symbol): Unit = {
         ownerNamer.enterInScope(f(ownerNamer.context.owner))
       }
       def addGetter(rtparams0: List[TypeDef])(create: (Namer, List[TypeDef]) => Tree): Unit = {

--- a/src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala
@@ -28,7 +28,7 @@ trait NamesDefaults { self: Analyzer =>
   // object, we need the templateNamer of that module class. These two are stored
   // as an attachment in the companion module symbol
   class ConstructorDefaultsAttachment(val classWithDefault: ClassDef, var companionModuleClassNamer: Namer) {
-    var defaults = mutable.ListBuffer[Symbol]()
+    val defaults = mutable.ListBuffer[Symbol]()
   }
 
   // Attached to the synthetic companion `apply` method symbol generated for case classes, holds

--- a/src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala
@@ -27,7 +27,9 @@ trait NamesDefaults { self: Analyzer =>
   // we need the ClassDef. To create and enter the symbols into the companion
   // object, we need the templateNamer of that module class. These two are stored
   // as an attachment in the companion module symbol
-  class ConstructorDefaultsAttachment(val classWithDefault: ClassDef, var companionModuleClassNamer: Namer)
+  class ConstructorDefaultsAttachment(val classWithDefault: ClassDef, var companionModuleClassNamer: Namer) {
+    var defaults = mutable.ListBuffer[Symbol]()
+  }
 
   // Attached to the synthetic companion `apply` method symbol generated for case classes, holds
   // the set contains all default getters for that method. If the synthetic `apply` is unlinked in

--- a/test/junit/scala/tools/nsc/typechecker/NamerTest.scala
+++ b/test/junit/scala/tools/nsc/typechecker/NamerTest.scala
@@ -18,6 +18,6 @@ class NamerTest extends BytecodeTesting {
     compiler.compileClasses("package p1; class Test { C.b(); C.a() }; object C { def a(x: Int = 0) = 0; def b(x: Int = 0) = 0 }")
     val methods = compiler.global.rootMirror.getRequiredModule("p1.C").info.decls.toList.map(_.name.toString).filter(_.matches("""(a|b).*"""))
     def getterName(s: String) = nme.defaultGetterName(TermName(s), 1).toString
-    Assert.assertEquals(List("a", "b", getterName("b"), getterName("a")), methods) // order depends on order of lazy type completion :(
+    Assert.assertEquals(List("a", getterName("a"), "b", getterName("b")), methods) // order depends on order of lazy type completion :(
   }
 }

--- a/test/junit/scala/tools/nsc/typechecker/NamerTest.scala
+++ b/test/junit/scala/tools/nsc/typechecker/NamerTest.scala
@@ -18,6 +18,6 @@ class NamerTest extends BytecodeTesting {
     compiler.compileClasses("package p1; class Test { C.b(); C.a() }; object C { def a(x: Int = 0) = 0; def b(x: Int = 0) = 0 }")
     val methods = compiler.global.rootMirror.getRequiredModule("p1.C").info.decls.toList.map(_.name.toString).filter(_.matches("""(a|b).*"""))
     def getterName(s: String) = nme.defaultGetterName(TermName(s), 1).toString
-    Assert.assertEquals(List("a", getterName("a"), "b", getterName("b")), methods) // order depends on order of lazy type completion :(
+    Assert.assertEquals(List("a", getterName("a"), "b", getterName("b")), methods) // order no longer depends on order of lazy type completion :)
   }
 }

